Added distance_threshold parameter to hierarchical clustering

VathsalaAchar opened the pull request:
#### Reference Issue 
Fixes #3796 

#### What does this implement/fix? Explain your changes.

Hierarchical clustering now has a `distance_threshold` parameter which can be set instead of `n_clusters` and this is used to determine the number of clusters to cut the tree at.

*Note: that this works only when compute_full_tree=True and n_clusters=None.*

Changes
* Either `distance_threshold` or `n_clusters` is accepted as parameter. When `distance_threshold` is set  `n_clusters` needs to be set to None as the default of 2 clusters has been retained.
* When building the tree the return_distance is set to True if the `distance_threshold` has been set.
The distances returned is then used to calculate the number of clusters when cutting the tree.

Tests
* Updated distance_threshold tests to pass with `n_clusters=None` as parameter as the default for n_clusters is 2.
* Test to raise error when neither n_cluster nor distance_threshold is passed in.
* Test agglomerative clustering with distance_threshold passed in and compare with the different number of clusters produced with and without connectivity.

Documentation
* Updated the User Guide to show `distance_threshold` parameter in the table.
* Doc string updated with `distance_threshold` explanation

#### Any other comments?

Is this implementation the right behaviour of `distance_threshold` or should the construction of the tree stop when the distance is reached?

--------
Apologies for the constant stream of commits, I was trying to sort out the errors myself.

VathsalaAchar committed:
Distance threshold added to hierarchical clustering

When distance_threshold is set then it is used to determine the number of clusters to cut the tree at.
Though it is to be noted that this works only when computer_full_tree=True.

* When building the tree the return_distance set to True if the distance_threshold has been set.
The distances returned is then used to calculate the number of clusters when cutting the tree.

* Test agglomerative clustering with distance_threshold passed in and compare with the different number of clusters produced with and without connectivity.

Changes to documentation to include distance_threshold

Updates to distance threshold in hierarchical clustering

* Moved the parameter check from init to fit for consistency

* Updates to tests to account for changes made above

Documentation changes based on review

* backticks for variables in docstrings

* formatting without backslashes

Test for hierarchical clustering with distance_threshold

* clusters produced are checked against the linkage tree to confirm that it matches the point where the distance exceeds the threshold set
* boundary case test when distance_threshold is equal to the distance

* Updated tests to compare clusters and number of clusters

* Allowing users to set n_clusters or distance_threshold and updated tests

* Checking the n_clusters None condition better

* Removed the necessity for n_clusters_ to be set to None and redundant checks

* Updated tests after the above changes

* Cleaned up test to compare clusters produced using n_clusters against distance_threshold

* Added and Simplified test for boundary conditions

* Updated the documentation on distance_threshold restrictions

massich commented on doc/modules/clustering.rst:
```diff
@@ -73,13 +73,13 @@ Overview of clustering methods
 
    * - :ref:`Ward hierarchical clustering <hierarchical_clustering>`
-     - number of clusters
+     - number of clusters or distance_threshold
```
remove the underscore (as distance threshold)

massich commented on doc/modules/clustering.rst:
```diff
@@ -73,13 +73,13 @@ Overview of clustering methods
 
    * - :ref:`Agglomerative clustering <hierarchical_clustering>`
-     - number of clusters, linkage type, distance
+     - number of clusters or distance_threshold, linkage type, distance
```
same here

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -595,6 +609,11 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
     ----------
     n_clusters : int, default=2
         The number of clusters to find.
+        NOTE: You should set either n_cluster or distance_threshold, NOT both.
```
variables in docstrings of public classes or functions should be surrounded by double back-ticks. 
i.e: \`\`n_clusters\`\` to render as ``n_clusters``

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -595,6 +609,11 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
+
+    distance_threshold : int (optional)
+        The distance threshold to cluster at.
+        NOTE: You should set either n_cluster or distance_threshold, NOT both.
```
same (missing backticks)

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -765,6 +806,11 @@ class FeatureAgglomeration(AgglomerativeClustering, AgglomerationTransform):
+
+    distance_threshold : int (optional)
+        The distance threshold to cluster at.
+        NOTE: You should set either n_cluster or distance_threshold, NOT both.
```
backticks

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -765,6 +806,11 @@ class FeatureAgglomeration(AgglomerativeClustering, AgglomerationTransform):
     ----------
     n_clusters : int, default 2
         The number of clusters to find.
+        NOTE: You should set either n_cluster or distance_threshold, NOT both.
```
backticks

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -696,7 +717,15 @@ def fit(self, X, y=None):
+            raise ValueError("Either n_clusters (>0) or distance_threshold "
+                             "needs to be set, got n_clusters={} and "
+                             "distance_threshold={} instead. "
+                             "Please set n_clusters=None to continue.".format(
```
instad of asking the user to set `n_clusters=None` :
* Set `n_clusters=None` and `distance_threshold=None` by default
* in the `fit` stage, do a validation to check if both are `None`, in that case revert to the previous default behavior of `n_clusters=2`
* Update the docstrings to reflect such behaviour

cc: @jnothman is this ok?

VathsalaAchar replied:
That was how I first attempted it, but I get a few errors from estimator_checks

`
AssertionError: Estimator FeatureAgglomeration should not change or mutate  the parameter n_clusters from None to 2 during fit.
`

Do I have to try and get around this error? Or force the user to set `n_clusters` as None when using `distance_threshold`?

raghavrv replied:
Instead of setting the `n_clusters`, we can make it a private attribute at __init__ like

```py
self._n_clusters = n_clusters
```

and have a property to define it's actual value like

```py
@property
def n_clusters(self):
    if self._n_clusters is None and self.distance_threshold is None:
        return 2
    elif self.distance_threshold is None ^ self._n_clusters is None:
        return self._n_clusters
    else:
        raise AttributeError("Both n_clusters and distance_threshold parameters"
                             "were set during initialization.")
```

jnothman replied:
All you need to do is what you first tried, only using local variables to pass around `n_clusters`, or store a new `n_clusters_` attribute at fit time.

massich commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -477,7 +576,7 @@ def test_compute_full_tree():
     # When n_clusters is less, the full tree should be built
     # that is the number of merges should be n_samples - 1
-    agc = AgglomerativeClustering(n_clusters=2, connectivity=connectivity)
+    agc = AgglomerativeClustering(connectivity=connectivity)
```
why this change?

VathsalaAchar replied:
The default is 2 clusters, does this need to be put back in for visibility?

massich replied:
yes

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -736,14 +765,26 @@ def fit(self, X, y=None):
+        distance_threshold = self.distance_threshold
+        # if distance_threshold is set then distances is returned
+        if distance_threshold:
+            self.children_, self.n_components_, self.n_leaves_, parents, \
```
Can you format it differently without backslashes like

```py
   (ch, com, le,
    pa) = (
       memory.ca...
```

massich commented on sklearn/cluster/hierarchical.py:
```diff
@@ -552,12 +554,24 @@ def _hc_cut(n_clusters, children, n_leaves):
 
     """
+    # cut the tree when distance threshold is reached
+    if distance_threshold:
```
Maybe explain in the param docstring of `distance_threshold` that setting `distance_threshold` would implicitly mean setting `n_clusters` to this expression. 

what are your thoughts @agramfort, @jnothman

massich reviewed (commented):
I think that appveyor failures would be fixed if you rebase up on master. 

Thanks for the PR.

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -214,6 +214,104 @@ def test_agglomerative_clustering():
     assert_array_equal(clustering.labels_, clustering2.labels_)
 
 
+def test_agglomerative_clustering_with_distance_threshold():
```
I don't think there's good reason to duplicate the tests for this parameter.

jnothman reviewed (commented):
I'd rather see a test which checks the invariant that we want to hold:
* set a distance threshold
* check that the clusters produced match the point in the linkage tree where the threshold would have been exceeded
* test the boundary case of the threshold equalling the distance

Btw, is this always specified as an absolute distance? Might we want (or prefer) to specify it relative to the average or median pairwise distance, for instance?

VathsalaAchar commented:
@massich I rebased off master to fix the CircleCI errors and now Travis fails while looking for ``nose-timer``. Is there anything I am supposed to do to fix this?

massich commented:
The problem comes from the way you had fixed [this comment](https://github.com/scikit-learn/scikit-learn/pull/9069#discussion_r124531272). Surely, I didn't express myself properly. What I was proposing was to use some dummy variables to improve readability. 

```py
if distance_threshold:
      (ch, com, le, pa) = \
           memory.cache...
else
       (ch, com, le, pa) = \
           memory.cache...
self.children_ = ch
self.n_comp...
```
That's why travis failed. You can execute the PEP8 check locally:
```sh
bash ./build_tools/travis/flake8_diff.sh
```

jnothman commented:
I don't think that's an appropriate solution Raghav

On 29 Jun 2017 1:57 am, "(Venkat) Raghav, Rajagopalan" <
notifications@github.com> wrote:

> *@raghavrv* commented on this pull request.
> ------------------------------
>
> In sklearn/cluster/hierarchical.py
> <https://github.com/scikit-learn/scikit-learn/pull/9069#discussion_r124582204>
> :
>
> > @@ -696,7 +717,15 @@ def fit(self, X, y=None):
>                               " instance, got 'memory={!r}' instead.".format(
>                                   type(memory)))
>
> -        if self.n_clusters <= 0:
> +        if (self.distance_threshold and self.n_clusters) or \
> +                (not self.n_clusters and not self.distance_threshold):
> +            raise ValueError("Either n_clusters (>0) or distance_threshold "
> +                             "needs to be set, got n_clusters={} and "
> +                             "distance_threshold={} instead. "
> +                             "Please set n_clusters=None to continue.".format(
>
> Instead of setting the n_clusters, we can make it a private attribute at
> *init* like
>
> self._n_clusters = n_clusters
>
> and have a property to define it's actual value like
>
> @propertydef n_clusters(self):
>     if self._n_clusters is None and self.distance_threshold is None:
>         return 2
>     elif self.distance_threshold is None ^ self._n_clusters is None:
>         return self._n_clusters
>     else:
>         raise AttributeError("Both n_clusters and distance_threshold parameters"
>                              "were set during initialization.")
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9069#discussion_r124582204>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69MQpGlj3YV9W2RZ-9cammqXszvdks5sInf8gaJpZM4N0SK3>
> .
>

raghavrv commented:
Are you then okay with users having to explicitly set `n_clusters=None` when using `distance_threshold`?

raghavrv commented:
I'm unable to recollect which other class has something like this. But I feel we did come across this problem before.

VathsalaAchar commented:
> bash ./build_tools/travis/flake8_diff.sh

@massich Thanks for this, I usually forget to test flake locally and noticed the one too many spaces later. But the error I was talking about was a time out error in a [previous build](https://travis-ci.org/scikit-learn/scikit-learn/builds/247946880), that fixed itself on the next push and hasn't appeared since. I have also fixed the variables for better readability.

@jnothman I have updated the tests. But could I get a quick review to see if I missed anything?

@jnothman @raghavrv How can I move forward on setting or not setting n_clusters=None when using distance_threshold?

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -513,8 +514,116 @@ def test_agg_n_clusters():
+                         return_distance=True)
+        clusters_at_threshold = np.count_nonzero(
+            distances >= distance_threshold) + 1
+        assert_true(clusters_at_threshold == clusters_produced)
```
No, I meant you should checked that the clusters, not the number of clusters, match.

VathsalaAchar commented:
Could someone help me finish this off? I made all the necessary changes but a quick review would help. Thanks in advance!

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -603,6 +619,13 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
     ----------
     n_clusters : int, default=2
         The number of clusters to find.
+        NOTE: You should set either ``n_cluster`` or ``distance_threshold``,
```
n_cluster -> n_clusters

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -603,6 +619,13 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
+
+    distance_threshold : int (optional)
+        The distance threshold to cluster at.
+        NOTE: You should set either ``n_cluster`` or ``distance_threshold``,
```
n_cluster -> n_clusters

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -744,14 +781,30 @@ def fit(self, X, y=None):
-                                       **kwargs)
+        distance_threshold = self.distance_threshold
+        # if distance_threshold is set then distances is returned
+        if distance_threshold:
```
is not None

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -730,12 +767,12 @@ def fit(self, X, y=None):
         if self.connectivity is None:
             compute_full_tree = True
-        if compute_full_tree == 'auto':
+        if compute_full_tree == 'auto' and not self.distance_threshold:
```
is not None

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -541,13 +541,15 @@ def _average_linkage(*args, **kwargs):
     ----------
     n_clusters : int or ndarray
         The number of clusters to form.
+        NOTE: You should set either n_cluster or distance_threshold, NOT both.
```
n_cluster -> n_clusters

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
+    n_clus = 0
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("Either n_clusters (>0) or distance_threshold "
+           "needs to be set, got n_clusters={} and "
```
Why do you get this error, and not the previous, when n_clusters=0?

VathsalaAchar replied:
Because of the order of the checks, the first check is general and checks for None and 0. I did this because None computes to less than 0 in Python 2 and not in Python 3. 

I can swap them around and add a check for None when checking if `n_clusters <=0` and this should give the same message for both cases.

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
+    n_clus, dist_thresh = 10, 10
+    rng = np.random.RandomState(0)
+    X = rng.rand(20, 10)
+    agc = AgglomerativeClustering(n_clusters=n_clus,
```
If this is setting n_clusters and distance_threshold, shouldn't it raise an error?

I would think that if distance_threshold is provided, n_clusters_ should be set to the number of clusters, not None.

VathsalaAchar replied:
No because `n_clusters_` get set to None if both params are set, thus voiding the check for both parameters. 

But now that I think about it, I am not going do the check, if `distance_threshold` is set then I'll ignore `n_clusters`. This will get rid of having to set `n_clusters_=None` which isn't of any use except for the check.

jnothman replied:
I'm a bit undecided about whether we should silently ignore n_clusters rather than force the user to explicit set it to None. But if we take the approach that setting a distance_threshold means n_clusters is ignored, then we need to state that in the documentation of both parameters.

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -704,9 +729,21 @@ def fit(self, X, y=None):
                                  type(memory)))
 
-        if self.n_clusters <= 0:
+        # if distance threshold is set we set n_clusters to None
```
Why is this helpful?

VathsalaAchar replied:
This was to pass the check "if both n_cluster and distance_threshold are set then fail". n_clusters_ does not need to be set to None for any computation as it gets ignored.

VathsalaAchar replied:
How do I handle the case when both `n_clusters` and `distance_threshold` are set? With default `n_clusters=2`, when `distance_threshold` is set it fails.

I had tried to get around this earlier by changing the default `n_clusters` to `None`, but that fails other tests. This was a compromise, but I'm not sure.

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
+
+
+def test_agglomerative_clustering_with_distance_threshold_edge_case():
+    # test boundary case of distance_threshold matching the distance
```
This test is okay, but I think it would be simpler to read and understand if you just do something like:

```
    X = [[0], [1]]
    for threshold, y_true in [(0.5, ...), (1.0, ...), (1.5, ...)]:
        # TODO: perhaps also loop over linkage
        clusterer = AgglomerativeClustering(n_clusters=None, distance_threshold=threshold)
        y_pred = clusterer.fit_predict(X)
        assert_equal(1, adjusted_rand_score(y_true, y_pred))
```

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
+        # test clusters produced
+        assert_true(np.array_equiv(clusters_produced, clusters_at_threshold))
+
+        # test if cluster number is different without connectivity
```
can we test the presence or absence of connectivity in a loop instead, for readability?

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
+        children, n_components, n_leaves, parent, distances = \
+            tree_builder(X, connectivity=connectivity, n_clusters=None,
+                         return_distance=True)
+        clusters_at_threshold = _hc_cut(n_clusters=None, children=children,
```
please move this down to "test clusters produced". I also think this should be using the `n_clusters` parameter rather than the `distance_threshold` parameter, to effectively test equivalence.

jnothman reviewed (commented):
There are a few small things that surprised me, like the `n_clusters_` behaviour and a couple of the error cases, but this is otherwise looking good.

blackyang commented:
Great feature! When will this be merged?

asanakoy commented:
Also waiting for the merge

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -704,9 +734,17 @@ def fit(self, X, y=None):
-        if self.n_clusters <= 0:
+        self.n_clusters_ = self.n_clusters
+
+        if self.n_clusters_ is not None and self.n_clusters_ <= 0:
```
why aren't you just using `self.n_clusters` here? `self.n_clusters` should be the user's input, and should be used in validation. `n_clusters_` should be output to the user.

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -560,12 +563,29 @@ def _hc_cut(n_clusters, children, n_leaves):
     n_leaves : int
         Number of leaves of the tree.
 
+    distance_threshold : int (optional)
```
int -> float

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -773,6 +827,14 @@ class FeatureAgglomeration(AgglomerativeClustering, AgglomerationTransform):
+        NOT both.
+
+    distance_threshold : int (optional)
+        The distance threshold to cluster at.
```
Update to match above

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -560,12 +563,29 @@ def _hc_cut(n_clusters, children, n_leaves):
 
+    distance_threshold : int (optional)
+        The distance threshold to cluster at.
+        If ``distance_threshold`` is set then ``n_clusters`` will be the
```
It is confusing here to refer to the `n_clusters` parameter. I'd rather: "If ``distance_threshold`` is set then  `n_clusters` will be ignored. Instead, the number of clusters will be determined by when distances between clusters first exceed the `distance_threshold` during agglomeration."

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -603,6 +623,14 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
+        NOT both.
+
+    distance_threshold : int (optional)
+        The distance threshold to cluster at.
```
Update to match above

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -541,13 +541,16 @@ def _average_linkage(*args, **kwargs):
     n_clusters : int or ndarray
         The number of clusters to form.
+        NOTE: You should set either ``n_clusters`` or ``distance_threshold``,
+        NOT both.
```
This comment doesn't really make sense, if `distance_threshold` overrides `n_clusters`. I think it's fine to only have the comment under `distance_threshold` saying that it being set makes this ignored.

jnothman reviewed (commented):
I don't understand what you intend by the `n_clusters_` attribute. It does not appear to be documented, and it does not seem to differ from `self.n_clusters`. If this PR introduces a new `n_clusters_` attribute, its role should be to report the number of clusters automatically identified by `distance_threshold`. Otherwise, I see no need for a new attribute.

VathsalaAchar commented:
> I don't understand what you intend by the n_clusters_ attribute. It does not appear to be documented, and it does not seem to differ from self.n_clusters. If this PR introduces a new n_clusters_ attribute, its role should be to report the number of clusters automatically identified by distance_threshold. Otherwise, I see no need for a new attribute.

You're absolutely right. The `n_clusters_` attribute made sense initially when I started off and wanted n_clusters to be None and this attribute was the workaround. But after all the changes it doesn't have any purpose now so I'll fix that. 

I'll also take care of the remaining documentation issues.

VathsalaAchar committed:
Changes based on review

* Doc string updates for clear information

* Removed redundant attribute n_clusters_

* Fixed tests

* Changes to FeatureAgglomeration to include distance threshold

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -743,14 +775,30 @@ def fit(self, X, y=None):
-            self.labels_ = _hc_cut(self.n_clusters, self.children_,
-                                   self.n_leaves_)
+            if distance_threshold is not None:
+                self.labels_ = _hc_cut(self.n_clusters, self.children_,
```
why don't you just calculate n_clusters from distances here and avoid modifying `_hc_cut` (adding two parameters for the sake of 1 line of logic)?

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,11 +529,44 @@ def test_agg_n_clusters():
+    n_clus = -1
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("n_clusters should be an integer greater than 0."
+           " %s was provided." % str(agc.n_clusters))
```
Either use a loop to test both -1 and 0, or put -1 directly in the expected error message

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,11 +529,44 @@ def test_agg_n_clusters():
+
+
+def test_agg_n_cluster_and_distance_threshold():
+    # Test that when distance_threshold is set n_clusters_ is unchanged
```
This isn't the right comment anymore

jnothman reviewed (commented):
I think what we should do is to change n_clusters to have a default value of None, which means "2 if distance_threshold is unset", and then raise an error when fitting a model with both `n_clusters` and `distance_threshold` set.

Otherwise, this is looking good.

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -559,3 +594,51 @@ def increment(self, *args, **kwargs):
+            children, n_components, n_leaves, parent, distances = \
+                tree_builder(X, connectivity=conn, n_clusters=None,
+                             return_distance=True)
+            num_clusters_at_threshold = np.count_nonzero(
```
Perhaps we should test something more explicit, just to be sure that your logic here is correct, like check that in single linkage, the minimum within-cluster pairwise distance for each sample is under the threshold and the minimum out-of-cluster pairwise distance is greater.

VathsalaAchar replied:
Do you mean along the lines of [this](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cluster/tests/test_hierarchical.py#L326) test? 
I could use the same dataset to do a more explicit test.

jnothman replied:
I don't see how it relates to that test. I mean that for some `X` and some predicted `labels`:
```
D = pairwise_distances(X, metric=metric)
for i in range(len(X)):
    in_cluster_mask = labels == labels[i]
    max_in_cluster_distance = D[i, in_cluster_mask].max()
    min_out_cluster_distance = D[i, ~in_cluster_mask].min()
    # XXX: there should be equality on one of these conditions
    assert max_in_cluster_distance < threshold
    assert min_in_cluster_distance > threshold
```

VathsalaAchar replied:
Apologies for the delayed response, but as far as I understand the pairwise distance will give the distance between each point in `X` <b>not</b> the distance between the clusters as they join up. The distance between each cluster is in the `distances` matrix calculated using the `scipy.cluster.hierarchy.linkage` method.

So is there still a need to have an explicit test?

jnothman replied:
True. But surely a similar invariance could be constructed about the average distances with average linkage...? I've not thought about it too rigorously.

VathsalaAchar replied:
Although what I said is true only when `connectivity` is `None`. If a `connectivity` matrix is passed in then calculating the clusters would mean deciphering what is happening in the `ward_tree` and `linkage_tree` methods, and I'm not sure it's worth the effort...

>True. But surely a similar invariance could be constructed about the average distances with average linkage...? I've not thought about it too rigorously.

I suppose, but is there a need to do this? 
I'm really not sure how to do a more explicit test, so I'd really appreciate help with this.

VathsalaAchar committed:
Updates based on review

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -783,6 +765,10 @@ def fit(self, X, y=None):
                                            **kwargs)
+            # find number of clusters to cut the tree when distance threshold
+            # is reached
+            self.n_clusters = np.count_nonzero(
```
But we don't allow modifying constructor parameters during fit (though this is not yet in common tests). So either use a local variable, or, indeed, store the computed n_clusters in a new n_clusters_ attribute (which should then be documented)

VathsalaAchar replied:
I've put comments in the code, is there any other place it needs to be documented?

VathsalaAchar committed:
Updates based on comments

* Added a new attribute `n_clusters_` to hold clusters computed when distance threshold is set

jnothman commented:
n_clusters_ should be documented in the Attributes section of the class docstring

VathsalaAchar committed:
Documentation for new attribute

jnothman commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -527,13 +529,44 @@ def test_agg_n_clusters():
+    n_clus, dist_thresh = None, 10
+    rng = np.random.RandomState(0)
+    X = rng.rand(20, 10)
+    agc = AgglomerativeClustering(n_clusters=n_clus,
```
Ideally we should check that you get the same behaviour if nclus is an int

amueller commented:
untagging for 0.20. complain if you disagree.

jnothman reviewed (commented):
This should be quite quick to review, for whichever release.

adrinjalali commented:
@VathsalaAchar could you please rebase on master?

adrinjalali committed:
Merge remote-tracking branch 'upstream/master' into hierarchical_clustering_threshold

adrinjalali committed:
fix n_components

adrinjalali committed:
move parameter to the end of the list

adrinjalali committed:
add whats_new entry

adrinjalali committed:
minor fix on n_clusters_

adrinjalali committed:
fix tests

adrinjalali committed:
fix docstrings

adrinjalali committed:
minor fix

adrinjalali committed:
remove assert_true

adrinjalali committed:
remove unrelated change

adrinjalali committed:
add a more explicit test

adrinjalali commented:
this looks good to me, but since I've applied my reviews here myself since the PR seemed stalled, another review is appreciated, should be an easy one.

jnothman commented:
@thomasjpfan, feel like giving this once over?

adrinjalali committed:
merge upstream/master

NicolasHug commented on sklearn/cluster/hierarchical.py:
```diff
@@ -788,10 +804,14 @@ def fit(self, X, y=None):
+        if self.n_clusters is None and self.distance_threshold is None:
+            raise ValueError("n_clusters and distance_threshold cannot be "
+                             "both None.")
+
```
We also need to check that:

- n_clusters is None if distance_threshold is not None
- distance_threshold is None if n_clusters is not None
- compute_full_tree is True if distance_threshold is not None

along with the associated tests.

Please also update n_clusters docstring to

>    n_clusters : int or None, default=2

adrinjalali replied:
I think the decision was that the `n_clusters` is ignored if `distance_threshold` is given. So it doesn't have to explicitly set as `None`, and that's why the first two checks are not here in the code. I don't have a strong position on either case.

> compute_full_tree is True if distance_threshold is not None

Good point.

NicolasHug replied:
> I think the decision was that the n_clusters is ignored if distance_threshold is given

Ok. How about not forcing it to be None, but still forcing it to have its default value (2, I think)?

adrinjalali replied:
And what would we do if the default value is explicitly given by the user? Do we want to detect that as well?

NicolasHug replied:
well you can't

NicolasHug replied:
That's why I think forcing it to none is a better idea.
It seems to me that we usually raise error when 2 exclusive attributes are used simultaneously.

adrinjalali replied:
Ok, done.

NicolasHug commented on sklearn/cluster/hierarchical.py:
```diff
@@ -814,7 +834,7 @@ def fit(self, X, y=None):
         if self.connectivity is None:
             compute_full_tree = True
-        if compute_full_tree == 'auto':
+        if compute_full_tree == 'auto' and self.distance_threshold is None:
```
I think we need to remove this and explicitly force compute_full_tree to be True if distance_threshold is passed

NicolasHug reviewed (commented):
First round, haven't reviewed the tests yet

adrinjalali commented on sklearn/cluster/hierarchical.py:
```diff
@@ -711,8 +711,21 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
             ``pooling_func`` has been deprecated in 0.20 and will be removed
             in 0.22.
 
+    distance_threshold : float (optional)
```
I hadn't changed this because the whole docstring of this class is a mishmash of a bunch of different styles. Ideally I'd rather fix them all.

NicolasHug replied:
I'm not too worried about the style, but we should still indicate that the default is None IMO

adrinjalali committed:
merge upstream/master

adrinjalali committed:
remove sentence, add to FeatureAgglomeration

adrinjalali commented on sklearn/cluster/hierarchical.py:
```diff
@@ -828,14 +848,31 @@ def fit(self, X, y=None):
+
+        distance_threshold = self.distance_threshold
+        # if distance_threshold is set then distances is returned
+        if distance_threshold is not None:
```
Nice!

adrinjalali committed:
code style change

adrinjalali committed:
check compute_full_tree, and docstring fix

adrinjalali committed:
apply more comments

adrinjalali committed:
force only one of the parameters to be non-None

adrinjalali committed:
merge upstream/master

NicolasHug commented on sklearn/cluster/hierarchical.py:
```diff
@@ -711,8 +711,19 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
             in 0.22.
 
+    distance_threshold : float, optional (default=None)
+        The distance threshold to cluster at. If not ``None``, ``n_clusters``
```
I still think this needs a description. It isn't clear what this parameter does right now.

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+        assert min_out_cluster_distance >= threshold
+
+
+def test_agglomerative_clustering_with_distance_threshold_edge_case():
```
we could parametrize this ideally but not very important

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+                distance_threshold=threshold,
+                linkage=linkage)
+            y_pred = clusterer.fit_predict(X)
+            assert_equal(1, adjusted_rand_score(y_true, y_pred))
```
use `assert a == b` instead?

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -573,6 +574,21 @@ def test_agg_n_clusters():
 
 
+def test_agg_n_cluster_and_distance_threshold():
+    # Test that when distance_threshold is set that n_clusters is ignored
```
I don't think this test really makes sense anymore now that both attributes are exclusive

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+                                            children=children,
+                                            n_leaves=n_leaves)
+            assert np.array_equiv(clusters_produced,
+                                  clusters_at_threshold)
```
the portion above should be a single test on its own since it's not related to what comes below.

Please also parametrize with `linkage` and `connectivity` (no need to compute it if it's None)

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+        n_clusters=None,
+        distance_threshold=1,
+        linkage="single").fit(X)
+    assert len(np.unique(clustering.labels_)) == 10
```
use `n_clusters_` attributes instead, and compare to `n_samples`.

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+    rng = np.random.RandomState(0)
+    n_samples = 10
+    X = rng.randint(-3, 3, size=(n_samples, 3))
+    # this should result in all data in their own clusters
```
For this to be obvious, maybe assert that the pairwise distances in X are all > 1?

Since X is random there's no reason for this to happen.

Or more simply, use a bigger range instead of -3, 3 so that it's easier to believe

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+    assert len(np.unique(clustering.labels_)) == 10
+
+    # check the distances within the clusters and with other clusters
+    threshold = 2
```
```suggestion
    distance_threshold = 2
```

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+    D = pairwise_distances(X, metric="euclidean")
+    # to avoid taking the 0 diagonal in min()
+    np.fill_diagonal(D, np.inf)
+    for i in np.unique(labels):
```
```suggestion
    for label in np.unique(labels):
```

NicolasHug commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -600,6 +616,105 @@ def increment(self, *args, **kwargs):
+    np.fill_diagonal(D, np.inf)
+    for i in np.unique(labels):
+        in_cluster_mask = labels == i
+        max_in_cluster_distance = (D[in_cluster_mask][:, in_cluster_mask]
```
`max_in_cluster_distance`  is always `inf`. 

related: every sample is in its own cluster, I'm not sure that was the intention of this test.

adrinjalali replied:
nice catch!

NicolasHug reviewed (commented):
Tests need an update

adrinjalali committed:
Merge remote-tracking branch 'upstream/master' into hierarchical_clustering_threshold

adrinjalali committed:
improve docstrings

adrinjalali committed:
merge upstream/master

adrinjalali committed:
improve tests and apply Nicolas's comments

adrinjalali committed:
merge upstream/master

NicolasHug commented:
Thanks for the update Adrin.

Just merged, @jnothman this can get in 0.21 right? Else we'll need to udpate the `versionadded` parts

jnothman commented:
0.21 has been branched. Should we move this to 0.22 to keep things clean?

jnothman commented:
Sorry I didn't see the last comment. I'd be happy to keep it out of 0.21 to
keep the lines clear. But we can cherry pick it in if you'd rather.

bede commented:
This feature can't come soon enough.

NicolasHug commented:
Sorry Joel I was out for the last few days and didn't answer, I just saw that you cherry picked it. Thanks!
